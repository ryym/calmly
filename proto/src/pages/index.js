import { createElement } from '../calmly';
import { Hoge } from './Hoge';

// import { dirname } from 'path';
// import { fileURLToPath } from 'url';

// console.log('index', import.meta.url);

import someImg from '../img/tashi_kani.png';

// XXX: この方法だと dist にも index.client.js が出力されちゃうのでやっぱダメ。
// webpack にも依存するし。。 (でもこれは不可避？)
import clientJs from './index.client.js';

export const IndexPage = () => {
  return (
    <html>
      <head />
      <body>
        <h1>Index</h1>
        <Hoge name="world" />
        <a href="/about">about</a>
        <br />

        <img src={someImg} alt="たしかに" />
        <script src="---js---" />
      </body>
    </html>
  );
};

IndexPage.calmlyConfig = {
  onClient: clientJs,
};
