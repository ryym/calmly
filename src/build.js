const webpack = require('webpack');
const path = require('path');
const fs = require('fs');
const os = require('os');
const { promisify } = require('util');
const React = require('react');
const ReactDOMServer = require('react-dom/server');
const { CalmlyContext } = require('./react-context');
const { loadWebpackConfigs } = require('./webpack');
const { loadPageConfigs } = require('./pages');

const writeFile = promisify(fs.writeFile);
const readFile = promisify(fs.readFile);
const mkdtemp = promisify(fs.mkdtemp);

const build = async () => {
  const cwd = process.cwd();

  const webpackConfigs = loadWebpackConfigs({ cwd });
  const distPath = webpackConfigs.htmlConfig.output.path;

  const pagesRoot = path.join(cwd, 'src', 'pages');
  const pages = await loadPageConfigs(pagesRoot);

  if (pages.length === 0) {
    throw new Error('no entry pages found');
  }

  const htmlEntries = pages.reduce((es, p) => {
    es[p.name] = path.join(pagesRoot, p.relativePath);
    return es;
  }, {});

  const rootPagesConfigPath = path.join(cwd, 'src', 'pages.js');
  const rootPagesConfigExists = fs.existsSync(rootPagesConfigPath);
  if (rootPagesConfigExists) {
    htmlEntries.pages = rootPagesConfigPath;
  }

  await runWebpack({
    ...webpackConfigs.htmlConfig,
    entry: htmlEntries,
  });

  const htmlManifest = require(path.join(distPath, 'manifest.json'));

  let renderHTML = undefined;
  if (rootPagesConfigExists) {
    const rootConfig = require(path.join(distPath, 'pages.js'));
    renderHTML = rootConfig.renderHTML;
  }

  const jsData = [];
  const renderResults = pages.map(page => {
    const render = domTree => {
      const ctxState = { paths: [] };
      const wrappedTree = React.createElement(
        CalmlyContext.Provider,
        { value: ctxState },
        domTree
      );
      const html = ReactDOMServer.renderToStaticMarkup(wrappedTree);
      jsData.push({ name: page.name, jsPaths: ctxState.paths });

      return new ResultHTML(page.name, html);
    };

    const outPath = path.join(distPath, page.relativePath);
    const rootComponent = require(outPath).default;
    const domTree = React.createElement(rootComponent, null);
    return renderHTML ? renderHTML(domTree, render) : render(domTree);
  });

  const tmpDir = await mkdtemp(path.join(os.tmpdir(), 'calmly-'));

  const jsResults = await Promise.all(
    jsData.map(async r => {
      let jsFilePath = null;

      if (r.jsPaths.length > 0) {
        let js = '';
        r.jsPaths.forEach((originalPath, i) => {
          js += `import _${i} from '${path.join(cwd, originalPath)}';`;
        });
        js += '\n';
        r.jsPaths.forEach((_path, i) => {
          js += `_${i}();`;
        });

        jsFilePath = path.join(tmpDir, `${r.name}.js`);
        await writeFile(jsFilePath, js);
      }

      return { name: r.name, jsFilePath, jsName: `${r.name}.js` };
    })
  );

  // TODO: Handle the case there are no client side JS.
  // (If entries is empty webpack throws an error)
  const entries = jsResults.reduce((es, r) => {
    if (r.jsFilePath) {
      es[r.name] = r.jsFilePath;
    }
    return es;
  }, {});

  await runWebpack({
    entry: entries,
    ...webpackConfigs.jsConfig,
  });

  const jsManifest = require(path.join(distPath, 'manifest.json'));

  await Promise.all(
    renderResults.map(html => {
      const jsResult = jsResults.find(r => r.name === html.name());

      if (jsResult.jsFilePath == null) {
        html.replace('scriptTag', '');
      } else {
        const realPath = jsManifest[jsResult.jsName];
        if (realPath == null) {
          throw new Error(
            `could not find JS file path for ${jsResult.jsName}. Something goes wrong.`
          );
        }
        const scriptTag = `<script src="/${realPath}"></script>`;
        html.replace('scriptTag', scriptTag);
      }

      const cssName = `${html.name()}.css`;
      const cssRealPath = htmlManifest[cssName];
      if (cssRealPath == null) {
        html.replace('stylesheetTag', '');
      } else {
        html.replace('stylesheetTag', `<link rel="stylesheet" href="${cssRealPath}" />`);
      }

      return writeFile(path.join(distPath, html.fileName()), html.toString());
    })
  );
};

const runWebpack = config => {
  return new Promise((resolve, reject) => {
    const compiler = webpack(config);
    compiler.run((err, stats) => {
      if (err) {
        reject(err);
      } else if (stats.hasErrors() || stats.hasWarnings()) {
        console.error(stats.toString({ chunks: false, colors: true }));
        reject(new Error('build error'));
      } else {
        resolve(stats);
      }
    });
  });
};

class ResultHTML {
  constructor(name, html) {
    this._name = name;
    this._html = html;
    this._replacements = [];
  }

  name() {
    return this._name;
  }

  fileName() {
    return `${this._name}.html`;
  }

  replace(key, value) {
    this._replacements.push({ key, value });
  }

  toString() {
    const html = this._replacements.reduce((s, r) => {
      return s.replace(`<style>#${r.key}{}</style>`, r.value);
    }, this._html);
    return html;
  }
}
module.exports = { build };
