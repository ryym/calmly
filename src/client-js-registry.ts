import { useContext, createContext, createElement } from 'react';

export interface ClientJSEntries {
  paths: string[];
}

export interface Registerer {
  register(path: string): void;
}

const Context = createContext<Registerer>({
  register() {
    throw new Error('this is dummy registerer');
  },
});

// XXX: It is nice if the path can be relative from the component file.
export const useClientJS = (filePath: string) => {
  const registerer = useContext(Context);
  registerer.register(filePath);
};

export class ClientJSRegistry {
  private paths: string[] = [];
  private setupDone = false;

  setupRegistration(rootDom: any): any {
    this.setupDone = true;
    const registerer = {
      register: (path: string) => {
        this.paths.push(path);
      },
    };
    return createElement(Context.Provider, { value: registerer }, rootDom);
  }

  getScriptFilePaths(): string[] | null {
    return this.setupDone ? this.paths : null;
  }
}
