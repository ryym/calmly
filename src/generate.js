const webpack = require('webpack');
const WebpackManifestPlugin = require('webpack-manifest-plugin');
const path = require('path');
const fs = require('fs');
const os = require('os');
const { promisify } = require('util');

const React = require('react');
const ReactDOMServer = require('react-dom/server');
const { CalmlyContext } = require('./react');

const writeFile = promisify(fs.writeFile);
const readFile = promisify(fs.readFile);
const mkdtemp = promisify(fs.mkdtemp);

const PLACEHOLDER_JS_PATH = '---js---';

const generate = async () => {
  const cwd = process.cwd();
  const distPath = path.join(cwd, 'dist');

  await runWebpack({
    mode: 'development',
    entry: path.join(cwd, 'src'),
    output: {
      path: distPath,
      libraryTarget: 'commonjs2',
      filename: 'index.js',
    },
    externals: 'calmly',
    module: {
      rules: [
        {
          test: /\.js$/,
          use: ['babel-loader'],
        },
      ],
    },
  });

  const { pages } = require(distPath);

  if (pages == null) {
    throw new Error('pages does not be exported');
  }

  const tmpDir = await mkdtemp(path.join(os.tmpdir(), 'calmly-'));

  const results = await Promise.all(
    Object.keys(pages).map(async name => {
      const domTree = pages[name]({ scriptUrl: PLACEHOLDER_JS_PATH });

      const ctxState = { paths: [] };
      const wrappedTree = React.createElement(
        CalmlyContext.Provider,
        { value: ctxState },
        domTree
      );
      const html = ReactDOMServer.renderToStaticMarkup(wrappedTree);

      if (ctxState.paths.length === 0) {
        return { name, html };
      }

      let js = '';
      ctxState.paths.forEach((originalPath, i) => {
        js += `import _${i} from '${path.join(cwd, originalPath)}';`;
      });
      js += '\n';
      ctxState.paths.forEach((_path, i) => {
        js += `_${i}();`;
      });

      const jsFilePath = path.join(tmpDir, `${name}.js`);
      await writeFile(jsFilePath, js);

      return { name, html, jsFilePath, jsName: `${name}.js` };
    })
  );

  const entries = results.reduce((es, r) => {
    if (r.jsFilePath) {
      es[r.name] = r.jsFilePath;
    }
    return es;
  }, {});

  await runWebpack({
    mode: 'development',
    entry: entries,
    output: {
      path: distPath,
    },
    module: {
      rules: [
        {
          test: /\.js$/,
          use: ['babel-loader'],
        },
      ],
    },
    plugins: [new WebpackManifestPlugin()],
  });

  const manifestJson = await readFile(path.join(distPath, 'manifest.json'), {
    encoding: 'utf8',
  });
  const manifest = JSON.parse(manifestJson);

  await Promise.all(
    results.map(r => {
      const realPath = manifest[r.jsName];
      const html = r.html.replace(PLACEHOLDER_JS_PATH, `/${realPath}`);
      return writeFile(path.join(distPath, `${r.name}.html`), html);
    })
  );
};

const runWebpack = config => {
  return new Promise((resolve, reject) => {
    const compiler = webpack(config);
    compiler.run((err, stats) => {
      if (err) {
        reject(err);
      } else if (stats.hasErrors() || stats.hasWarnings()) {
        console.error(stats.toString({ chunks: false, colors: true }));
        reject(new Error('build error'));
      } else {
        resolve(stats);
      }
    });
  });
};

module.exports = { generate };
