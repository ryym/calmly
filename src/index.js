const { generate } = require('./generate');
const { useClientJS } = require('./react');

module.exports = { generate, useClientJS };
