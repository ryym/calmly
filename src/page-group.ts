export class PageGroup {
  constructor(readonly name: string, private readonly templates: NamedPageTemplate[]) {
    if (templates.length === 0) {
      throw new Error('PageGroup templates must not be empty');
    }
  }

  get clientJsPaths(): string[] {
    // All templates shares same client scripts.
    return this.templates[0].template.clientJsPaths;
  }

  replace(key: string, value: string) {
    this.templates.forEach(t => t.template.replace(key, value));
  }

  renderPages(): Page[] {
    return this.templates.map(t => {
      const html = t.template.render();
      return { name: t.name, html };
    });
  }
}

export interface Page {
  readonly name: string;
  readonly html: string;
}

export interface NamedPageTemplate {
  readonly name: string;
  readonly template: PageTemplate;
}

export class PageTemplate {
  private readonly replacements: { key: string; value: string }[];

  constructor(private readonly html: string, readonly clientJsPaths: string[]) {
    this.replacements = [];
  }

  replace(key: string, value: string) {
    this.replacements.push({ key, value });
  }

  render(): string {
    const html = this.replacements.reduce((s, r) => {
      // TODO: Do more practical and secure replacement..
      return s.replace(`<style>#${r.key}{}</style>`, r.value);
    }, this.html);
    return html;
  }
}
