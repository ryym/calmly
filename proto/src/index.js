import { IndexPage } from './pages/index';
import { AboutPage } from './pages/about';
import { createElement } from '../calmly';

export const pages = {
  index: () => <IndexPage />,
  about: () => <AboutPage />,
};
