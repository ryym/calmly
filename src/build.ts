import webpack from 'webpack';
import * as path from 'path';
import * as fs from 'fs';
import * as os from 'os';
import { promisify } from 'util';
import * as React from 'react';
import * as ReactDOMServer from 'react-dom/server';
import { loadWebpackConfigs } from './webpack';
import { loadRoutes, Route } from './routes';
import { ClientJSRegistry } from './client-js-registry';
import { PH_SCRIPT_TAG, PH_STYLESHEET_TAG } from './placeholder';
import { PageGroup, PageTemplate } from './page-group';

const writeFile = promisify(fs.writeFile);
const mkdir = promisify(fs.mkdir);
const mkdtemp = promisify(fs.mkdtemp);
const removeFile = promisify(fs.unlink);

const stripExtension = (filePath: string): string => {
  // TODO: Support other extensions.
  return filePath.replace(/\.js$/, '');
};

export interface BuildOptions {
  readonly cwd?: string;
}

export const build = async (opts: BuildOptions = {}) => {
  const cwd = opts.cwd || process.cwd();

  if (!path.isAbsolute(cwd)) {
    // This is because we need to 'require()' the config file using the cwd.
    throw new Error('cwd must be absolute path');
  }

  const webpackConfigs = loadWebpackConfigs({ cwd });
  const distPath = webpackConfigs.htmlConfig.output.path;

  const routesRoot = path.join(cwd, 'src', 'pages');
  const routes = await loadRoutes(routesRoot);

  if (routes.length === 0) {
    throw new Error('no entry pages found');
  }

  const pageManifest = await buildPageEntries(routesRoot, routes, {
    webpack: webpackConfigs.htmlConfig,
    distPath,
  });

  const pageGroups = await Promise.all(routes.map((route) => renderPages(route, distPath)));

  const clientJSEntryFiles = await writeClientJSEntryFiles(pageGroups, cwd);
  const jsManifest = await buildClientJSEntries(clientJSEntryFiles, {
    webpack: webpackConfigs.jsConfig,
    distPath,
  });

  await Promise.all(
    pageGroups.map(async (pg) => {
      const clientJSEntry = clientJSEntryFiles.find((r) => r.pageName === pg.name);

      // Replace script tag placeholders.
      if (clientJSEntry == null) {
        pg.replace(PH_SCRIPT_TAG, '');
      } else {
        const realPath = jsManifest[clientJSEntry.jsName];
        const scriptTag = `<script src="/${realPath}"></script>`;
        pg.replace(PH_SCRIPT_TAG, scriptTag);
      }

      // Replace stylesheet tag placeholders.
      const cssName = `${pg.name}.css`;
      const cssRealPath = pageManifest[cssName];
      pg.replace(
        PH_STYLESHEET_TAG,
        cssRealPath == null ? '' : `<link rel="stylesheet" href="${cssRealPath}" />`
      );

      await Promise.all(
        pg.renderPages().map(async (page) => {
          if (page.name.endsWith('index')) {
            await writeFile(path.join(distPath, `${page.name}.html`), page.html);
          } else {
            const pageDirPath = path.join(distPath, page.name);
            if (!fs.existsSync(pageDirPath)) {
              await mkdir(pageDirPath);
            }
            await writeFile(path.join(distPath, page.name, 'index.html'), page.html);
          }
        })
      );
    })
  );

  await removeComponentFiles(distPath, routes);
};

const runWebpack = (config: any) => {
  return new Promise((resolve, reject) => {
    const compiler = webpack(config);
    compiler.run((err, stats) => {
      if (err) {
        reject(err);
      } else if (stats.hasErrors() || stats.hasWarnings()) {
        console.error(stats.toString({ chunks: false, colors: true }));
        reject(new Error('build error'));
      } else {
        resolve(stats);
      }
    });
  });
};

interface RouteConfig {
  renderHTML(domTree: any, render: (domTree: any) => PageTemplate): PageTemplate;
}

const writeClientJSEntryFiles = async (
  pageGroups: PageGroup[],
  cwd: string
): Promise<ClientJSEntryFile[]> => {
  const tmpDir = await mkdtemp(path.join(os.tmpdir(), 'calmly-'));
  const jsPageGroups = pageGroups.filter((p) => p.clientJsPaths.length > 0);

  return await Promise.all(
    jsPageGroups.map(async (pg) => {
      const jsPaths = pg.clientJsPaths;
      let js = '';
      jsPaths.forEach((originalPath, i) => {
        js += `import _${i} from '${path.join(cwd, originalPath)}';`;
      });
      js += '\n';
      jsPaths.forEach((_path, i) => {
        js += `_${i}();`;
      });

      const entryName = `${pg.name}.client`;
      const jsName = `${pg.name}.client.js`;
      const filePath = path.join(tmpDir, jsName);
      await writeFile(filePath, js);
      return { pageName: pg.name, entryName, jsName, filePath };
    })
  );
};

interface ClientJSEntryFile {
  readonly pageName: string;
  readonly entryName: string;
  readonly jsName: string;
  readonly filePath: string;
}

interface BuildConfig {
  readonly distPath: string;
  readonly webpack: object;
}

type WebpackManifest = {
  readonly [path: string]: string;
};

const buildPageEntries = async (
  routesRoot: string,
  routes: Route[],
  config: BuildConfig
): Promise<WebpackManifest> => {
  const htmlEntries = routes.reduce<{ [key: string]: any }>((es, r) => {
    es[r.name] = path.join(routesRoot, r.filePath);
    if (r.configPath != null) {
      const entryName = stripExtension(r.configPath);
      if (es[entryName] == null) {
        es[entryName] = path.join(routesRoot, r.configPath);
      }
    }
    return es;
  }, {});

  await runWebpack({ ...config.webpack, entry: htmlEntries });
  return require(path.join(config.distPath, 'manifest.json'));
};

const buildClientJSEntries = async (
  clientJSEntryFiles: ClientJSEntryFile[],
  config: BuildConfig
): Promise<WebpackManifest> => {
  if (clientJSEntryFiles.length === 0) {
    return {};
  }

  const entries = clientJSEntryFiles.reduce<{ [key: string]: string }>((es, r) => {
    es[r.entryName] = r.filePath;
    return es;
  }, {});
  await runWebpack({ ...config.webpack, entry: entries });
  return require(path.join(config.distPath, 'js.manifest.json'));
};

const renderPages = async (route: Route, distPath: string): Promise<PageGroup> => {
  const render = (domTree: any) => {
    const jsRegistry = new ClientJSRegistry();
    const wrappedTree = jsRegistry.setupRegistration(domTree);
    const html = ReactDOMServer.renderToStaticMarkup(wrappedTree);
    const jsPaths = jsRegistry.getScriptFilePaths()!;
    return new PageTemplate(`<!DOCTYPE html>${html}`, jsPaths);
  };

  let routeConfig: RouteConfig | null = null;
  if (route.configPath) {
    const fullConfigPath = path.join(distPath, route.configPath);
    routeConfig = require(fullConfigPath);
  }

  const outPath = path.join(distPath, route.filePath);
  const { default: rootComponent, ...componentConfig } = require(outPath);

  if (route.isDynamic()) {
    const { getInitialPropsMap } = componentConfig;
    if (getInitialPropsMap == null) {
      throw new Error('dynamic page must be export getInitialPropsMap');
    }
    // TODO: Consider the case where propsMap is empty.
    const propsMap: Map<unknown, object> = await getInitialPropsMap();
    const templates = Array.from(propsMap.entries()).map(([name, props]) => {
      const domTree = React.createElement(rootComponent, props);
      const template = routeConfig ? routeConfig.renderHTML(domTree, render) : render(domTree);
      return { name: path.join(route.dirName(), String(name)), template };
    });
    return new PageGroup(route.name, templates);
  } else {
    const { getInitialProps } = componentConfig;
    const initialProps = getInitialProps ? await getInitialProps() : null;
    const domTree = React.createElement(rootComponent, initialProps);
    const template = routeConfig ? routeConfig.renderHTML(domTree, render) : render(domTree);

    return new PageGroup(route.name, [{ name: route.name, template }]);
  }
};

const removeComponentFiles = async (distPath: string, routes: Route[]) => {
  const uniqueConfigPaths = routes.reduce<Set<string>>((paths, route) => {
    if (route.configPath != null) {
      paths.add(route.configPath);
    }
    return paths;
  }, new Set());

  return Promise.all([
    ...Array.from(uniqueConfigPaths).map((configPath) =>
      removeFile(path.join(distPath, configPath))
    ),
    ...routes.map((route) => removeFile(path.join(distPath, route.filePath))),
  ]);
};
