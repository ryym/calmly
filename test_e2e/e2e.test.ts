import * as fs from 'fs';
import { promisify } from 'util';
import * as path from 'path';
import * as prettier from 'prettier';
import { build } from '../src/build';

const readdir = promisify(fs.readdir);
const fileStat = promisify(fs.stat);
const readFile = promisify(fs.readFile);

// This excludes directories.
const readdirRec = async (root: string, intermediate = ''): Promise<string[]> => {
  let allFiles: string[] = [];
  const dirPath = path.join(root, intermediate);
  const files = await readdir(dirPath);
  for (let fileName of files) {
    const relPath = path.join(intermediate, fileName);
    const stat = await fileStat(path.join(root, relPath));
    if (stat.isDirectory()) {
      const children = await readdirRec(root, relPath);
      allFiles = [...allFiles, ...children];
    } else {
      allFiles.push(relPath);
    }
  }
  return allFiles;
};

describe('E2E', () => {
  const e2eDir = path.join(__dirname, 'samples');
  const testDirs = fs.readdirSync(e2eDir);

  for (let testDirName of testDirs) {
    it(testDirName, async () => {
      const testDirPath = path.join(e2eDir, testDirName);
      await build({ cwd: testDirPath });

      const distPath = path.join(testDirPath, 'dist');
      const distFiles = await readdirRec(distPath);
      for (let filePath of distFiles) {
        const content = await readFile(path.join(distPath, filePath));
        const ext = path.extname(filePath);
        switch (ext) {
          case '.html': {
            const html = prettier.format(String(content), { parser: 'html' });
            expect(html).toMatchSnapshot(filePath);
            break;
          }

          case '.json': {
            const json = JSON.parse(String(content));
            expect(json).toMatchSnapshot(filePath);
            break;
          }

          default:
            throw new Error(`unexpected file extension: ${ext}`);
        }
      }
    });
  }
});
