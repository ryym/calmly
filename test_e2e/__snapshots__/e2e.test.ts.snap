// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`E2E helloworld: index.html 1`] = `
"<!DOCTYPE html>
<body>
  <h1 id=\\"hello\\">Hello</h1>
  <div>content</div>
</body>
"
`;

exports[`E2E helloworld: manifest.json 1`] = `
Object {
  "index.js": "index.js",
}
`;
